src/actions/plugins/targets/smartQuotesMatcher.ts

```diff
@@ -0,0 +1,336 @@
+enum QuoteMatch {
+  Opening,
+  Closing,
+}
+export type WhichQuotes = 'current' | 'next' | 'last';
+type Dir = '>' | '<';
+type SearchAction = {
+  first: Dir;
+  second: Dir;
+  includeCurrent: boolean;
+};
+type QuotesAction = {
+  search: SearchAction | undefined;
+  skipToLeft: number; // for last quotes, how many quotes need to skip while searching
+  skipToRight: number; // for next quotes, how many quotes need to skip while searching
+};
+
+const quoteDirs: Record<string, QuotesAction> = {
+  '002': {
+    // | "a" "b" "c"
+    search: { first: '>', second: '>', includeCurrent: false },
+    skipToLeft: 0,
+    skipToRight: 1,
+  },
+  '012': {
+    // |"a" "b" "c" "
+    search: { first: '>', second: '>', includeCurrent: true },
+    skipToLeft: 0,
+    skipToRight: 2,
+  },
+  '102': {
+    // "a" "|b" "c" "
+    search: { first: '<', second: '>', includeCurrent: false },
+    skipToLeft: 2,
+    skipToRight: 2,
+  },
+  '112': {
+    //  "a" "b|" "c"
+    search: { first: '<', second: '<', includeCurrent: true },
+    skipToLeft: 2,
+    skipToRight: 1,
+  },
+  '202': {
+    //  "a"| "b" "c"
+    search: { first: '>', second: '>', includeCurrent: false },
+    skipToLeft: 1,
+    skipToRight: 1,
+  },
+  '211': {
+    //  "a" |"b" "c"
+    search: { first: '>', second: '>', includeCurrent: true },
+    skipToLeft: 1,
+    skipToRight: 2,
+  },
+  '101': {
+    //  "a" "|b" "c"
+    search: { first: '<', second: '>', includeCurrent: false },
+    skipToLeft: 2,
+    skipToRight: 2,
+  },
+  '011': {
+    //  |"a" "b" "c"
+    search: { first: '>', second: '>', includeCurrent: true },
+    skipToLeft: 0,
+    skipToRight: 2,
+  },
+  '110': {
+    //  "a" "b" "c|"
+    search: { first: '<', second: '<', includeCurrent: true },
+    skipToLeft: 2,
+    skipToRight: 0,
+  },
+  '212': {
+    //  "a" |"b" "c" "
+    search: { first: '>', second: '>', includeCurrent: true },
+    skipToLeft: 1,
+    skipToRight: 2,
+  },
+  '111': {
+    //  "a" "b|" "c" "
+    search: { first: '<', second: '<', includeCurrent: true },
+    skipToLeft: 2,
+    skipToRight: 1,
+  },
+  '200': {
+    //  "a" "b" "c"|
+    search: { first: '<', second: '<', includeCurrent: false },
+    skipToLeft: 1,
+    skipToRight: 0,
+  },
+  '201': {
+    //  "a" "b" "c"| "
+    //  "a"| "b" "c" "
+    search: { first: '>', second: '>', includeCurrent: false },
+    skipToLeft: 1,
+    skipToRight: 1,
+  },
+  '210': {
+    //  "a" "b" "c" |"
+    search: undefined,
+    skipToLeft: 1,
+    skipToRight: 0,
+  },
+  '001': {
+    // | "a" "b" "c" "
+    search: undefined,
+    skipToLeft: 0,
+    skipToRight: 1,
+  },
+  '010': {
+    //  a|"b
+    search: undefined,
+    skipToLeft: 0,
+    skipToRight: 0,
+  },
+  '100': {
+    //  "a" "b" "c" "|
+    search: undefined,
+    skipToLeft: 2,
+    skipToRight: 0,
+  },
+  '000': {
+    //  |ab
+    search: undefined,
+    skipToLeft: 0,
+    skipToRight: 0,
+  },
+};
+
+export class SmartQuoteMatcher {
```

Comment:
Can you add a comment to this class describing how it differs from the usual `i"`?

Reply:
```suggestion
/**
 * An extended of `QuoteMatcher` (though not inheritance because it's too differ from the original).
 *
 * This class is intended to handle quotes-object in a "smarter" way than the default one. e.g., when
 * you have a line ` "a"| "b" ` and the cursor in between the quotes-objects, instead of treating the
 * space between as a quote-object, we will take "b" to be our object of interest.
 * So, in this case, `di"` would result in ` "a" "|" `, instead of ` "a""b" `.
 */
export class SmartQuoteMatcher {
```
Does that make sense?

Reply:
Yup! I think `extended` and `differ` are typos though

Reply:
updated. feel free to suggest an alternative description, though.